crates/ntx-builder/src/actor/account_state.rs

```diff
@@ -0,0 +1,348 @@
+    #[instrument(target = COMPONENT, name = "ntx.state.load", skip_all)]
+    pub async fn load(
+        account: Account,
+        account_prefix: NetworkAccountPrefix,
+        store: &StoreClient,
+        block_num: BlockNumber,
+    ) -> Result<Self, StoreError> {
+        let notes = store.get_unconsumed_network_notes(account_prefix, block_num.as_u32()).await?;
+        let notes = notes
+            .into_iter()
+            .filter_map(|note| {
+                if let NetworkNote::SingleTarget(note) = note {
+                    Some(note)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        let account = NetworkAccountNoteState::new(account, notes);
+
+        let state = Self {
+            account,
+            account_prefix,
+            inflight_txs: BTreeMap::default(),
+            nullifier_idx: HashSet::default(),
+        };
+
+        state.inject_telemetry();
+
+        Ok(state)
+    }
+
+    /// Selects the next candidate network transaction.
+    #[instrument(target = COMPONENT, name = "ntx.state.select_candidate", skip_all)]
+    pub fn select_candidate(
+        &mut self,
+        limit: NonZeroUsize,
+        chain_state: ChainState,
+    ) -> Option<TransactionCandidate> {
+        // Remove notes that have failed too many times.
+        self.account.drop_failing_notes(Self::MAX_NOTE_ATTEMPTS);
+
+        // Skip empty accounts, and prune them.
+        // This is how we keep the number of accounts bounded.
+        if self.account.is_empty() {
+            return None;
+        }
+
+        // Select notes from the account that can be consumed or are ready for a retry.
+        let notes = self
+            .account
+            .available_notes(&chain_state.chain_tip_header.block_num())
+            .take(limit.get())
+            .cloned()
+            .collect::<Vec<_>>();
+
+        // Skip accounts with no available notes.
+        if notes.is_empty() {
+            return None;
+        }
+
+        let (chain_tip_header, chain_mmr) = chain_state.into_parts();
+        TransactionCandidate {
+            account: self.account.latest_account(),
+            notes,
+            chain_tip_header,
+            chain_mmr,
+        }
+        .into()
+    }
+
+    /// Marks notes of a previously selected candidate as failed.
+    ///
+    /// Does not remove the candidate from the in-progress pool.
+    #[instrument(target = COMPONENT, name = "ntx.state.notes_failed", skip_all)]
+    pub fn notes_failed(&mut self, notes: &[Note], block_num: BlockNumber) {
+        let nullifiers = notes.iter().map(Note::nullifier).collect::<Vec<_>>();
+        self.account.fail_notes(nullifiers.as_slice(), block_num);
+    }
+
+    /// Updates state with the mempool event.
+    #[instrument(target = COMPONENT, name = "ntx.state.mempool_update", skip_all)]
+    pub async fn mempool_update(&mut self, update: MempoolEvent) -> Option<ActorShutdownReason> {
```

Comment:
Does this function need to be `async`? It does not contain any `await`.

Reply:
Thanks, idk why clippy is missing these now.